atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStore.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Gets the upper timestamp limit from the database if it exists.
+     * @return Timestamp limit stored in the KVS, if present
+     * @throws IllegalStateException if the timestamp limit has been invalidated
+     */
+    public synchronized Optional<Long> getUpperLimit() {
+        return clientPool().runWithRetry(client -> {
+            BoundData boundData = getCurrentBoundData(client);
+            if (boundData.bound() == null) {
+                return Optional.<Long>empty();
+            }
+            Preconditions.checkState(CassandraTimestampUtils.isValidTimestampData(boundData.bound()),
+                    "The timestamp bound cannot be read.");
+            return Optional.of(PtBytes.toLong(boundData.bound()));
+        });
+    }
+
+    /**
+     * Stores an upper timestamp limit in the database in an atomic way.
+     * @param expected Expected current value of the timestamp bound (or null if there is no bound)
+     * @param target Upper timestamp limit we want to store in the database
+     * @throws ConcurrentModificationException if the expected value does not match the bound in the database
+     */
+    public synchronized void storeTimestampBound(Optional<Long> expected, long target) {
+        clientPool().runWithRetry(client -> {
+            byte[] expectedBytes = expected.map(PtBytes::toBytes).orElse(null);
+            byte[] targetBytes = PtBytes.toBytes(target);
+            ByteBuffer queryBuffer = CassandraTimestampUtils.constructCheckAndSetMultipleQuery(
+                    ImmutableMap.of(
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            Pair.create(expectedBytes, targetBytes)));
+            CqlResult result = executeQueryUnchecked(client, queryBuffer);
+            checkOperationWasApplied(expected, target, result);
+            return null;
+        });
+    }
+
+    /**
+     * Writes a backup of the existing timestamp to the database, if none exists. After this backup, this timestamp
+     * service can no longer be used until a restore. Note that the value returned is the value that was backed up;
+     * multiple calls to this method are safe, and will return the backed up value.
+     *
+     * @param defaultValue value to backup if the timestamp table is empty
+     * @return value of the timestamp that was backed up, if applicable
+     */
+    public synchronized long backupExistingTimestamp(long defaultValue) {
+        return clientPool().runWithRetry(client -> {
+            BoundData boundData = getCurrentBoundData(client);
+            byte[] currentBound = boundData.bound();
+            byte[] currentBackupBound = boundData.backupBound();
+
+            if (CassandraTimestampUtils.isValidTimestampData(currentBackupBound)) {
+                // Backup bound has been updated!
+                Preconditions.checkState(!CassandraTimestampUtils.isValidTimestampData(currentBound),
```

Comment:
I guess here you do have to check the validity of the data. We can either use CassandraTimestampBoundStoreEntry to determine it, which is probably better, or alternatively just check for 1 bit as a failure and return true otherwise. That would make it more generic, but also susceptible to legitimate corruption. In any case, for this PR, this is good enough